### PR TITLE
Updating image.py for safer datatype conversion

### PR DESCRIPTION
**Pull request by mittimithai**
(apologies, I created a new pull request from [4772](https://github.com/astropy/astropy/pull/4772)
as I couldn't figure out who to rebase the old one as I made it from editing the sourcefile on github..my github-fu is unfortunately weak, I've adjusted the latest comment from that pull request below)

My solution indeed may not be correct after tracing through the code(though it appears to work). The error is triggered from lxnstack when opens a canon raw CR2 file and tries to convert it to FITS with astropy.

I put the the longer trace below, lnxnstack fills 'data' with type 'uint16' and then eventually applies 'scale' with a destination data type of 'int16'. The add happens before the conversion of 'data' to the new type. The conversion should happen before zeroing and scaling, the zero should be constructed to be the same type as 'data'. 

I think this change fixes the problem more correctly than the existing version.

```
File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/lxnstack/utils.py", line 1953, in imwrite_fits
outbits)
File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/lxnstack/utils.py", line 1983, in _get_fits_hdl
hdu.scale('int16', bzero=32768)
File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/astropy/io/fits/hdu/image.py", line 384, in scale
bzero=bzero, blank=None)
File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/astropy/io/fits/hdu/image.py", line 443, in _scale_internal
self.data += -_zero
TypeError: Cannot cast ufunc add output from dtype('int32') to dtype('uint16') with casting rule 'same_kind'
```

### CODE REVIEW

**Comment by pllim** [comment]
@mittimithai , as far as I can understand from your description, it looks like the change should actually be in `lnxnstack` on how it manages its data types, not in Astropy. It does not make sense to me that it would define `uint` type and then attempts to apply `int` to the data...

**Comment by mittimithai** [comment]
I haven't gone into code in detail to understand exactly when data should be signed/unsigned, but it isn't obvious to me that lxnstack is the problem here.

What's wrong with trying scale uint16 data scale to int16 data with a positive zero?

**Comment by pllim** [comment]
I don't think this PR is acceptable at this point because lots of tests failed.

**Comment by mittimithai** [comment]
Yes, I can see now that some of the NaN handling seems to fail.  I'll try and fix, but you do agree that fundamentally we should be able to scale uint16 to int16, yes?

**Comment by pllim** [comment]
@mittimithai , can you just use the following code without changing Astropy? Will that work for you?

```python
>>> hdu.data = hdu.data.astype('int16')
>>> hdu.scale(bzero=32768)
```

**Comment by mittimithai** [comment]
I can test this in lxnstack, but the scale function should put some constraints (if only in docstring) on source destination types. Still isn't obvious to me why we should not expect this function to work with int16 and uint16.

**Comment by pllim** [comment]
I am not familiar enough with FITS standards to answer that question. But just looking at the code, which cast `-` to the zeropoint, thus changing it to a negative number, it seems that the logic is not intended to be used on unsigned int at all.

**Comment by mittimithai** [comment]
Hmm, I still have a lot of failures in testing when I tried (what I thought) would be the most conservative and simple solution:

replacing:
self.data += -_zero
with:
 self.data = self.data - _zero

This would not trigger the numpy error and use the standard casting rules.

**Comment by mittimithai** [comment]
Ok now I am confused, I reverted my change in this fork and the checks keep failing.  It isn't clear to me that my (tiny) change is what made stuff fail.  Apologies if my noob is showing here.

File: astropy/io/fits/hdu/image.py [hunk 1]
```diff
@@ -439,7 +439,8 @@ def _scale_internal(self, type=None, option='old', bscale=1, bzero=0,
         # Do the scaling
         if _zero != 0:
             # 0.9.6.3 to avoid out of range error for BZERO = +32768
-            self.data += -_zero
+#            self.data = self.data - _zero
+            self.data+=-zero
```
Comment (pllim) [inline astropy/io/fits/hdu/image.py]: Typo here. The variable name is `_zero`, not `zero`.

**Comment by mittimithai** [comment]
Finally, that took embarrassingly long.  I am pretty sure this is the right solution, it marks the area clearly and keep it compatible with newer numpy versions.  No typos and passes regression tests.

**Comment by pllim** [comment]
Thanks, @mittimithai . This looks a lot better! @embray , are you okay with this?

**Comment by astrofrog** [comment]
@mittimithai - given that the changes are only a couple of lines, could you rebase and squash all commits into a single commit?

**Comment by astrofrog** [comment]
I've rebased this and squashed the commits in https://github.com/astropy/astropy/pull/5250

**Pull request closed without merging**